File: go/classic/removeDuplicates_80.go
package classic

func removeDuplicates(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return -1
	}
	// i, j 代表一对相对的区间的开始和结尾的index
	// j 指向了下一个不等于的地方
	i, j := 0, 0
	// 更新的位置
	counter := 0
	for j < len(nums) {
		for j < len(nums) && nums[j] == nums[i] {
			j++
		}
		// 赋值一片
		for z := 0; z < min(2, j-i); z++ {
			nums[counter] = nums[i]
			counter++
		}
		i = j
	}
	return counter
}

File: java/src/RemoveDuplicates_80.java
public class RemoveDuplicates_80 {

    public int removeDuplicates(int[] nums) {
        // 删除排序数组的重复项 能够保留两个重复项

        // l、r 框定一个区域
        int l = 0, r = 0;
        // 最后返回结果的 index
        int newIndex = 0;
        while (r < nums.length) {
            while (r < nums.length && nums[r] == nums[l]) {
                r++;
            }

            for (int i = 0; i < Math.min(2, r - l); i++) {
                nums[newIndex++] = nums[l];
            }
            l = r;
        }

        return newIndex;
    }
}
